File: bikedata.py
import csv
import numpy as np
from scipy import signal
from matplotlib import *
import matplotlib.pyplot as plt
import serverAnalysis
from mpl_toolkits.mplot3d import Axes3D
from matplotlib.collections import PolyCollection
import os

"""
Digital Gym Research Group - Ashu Sabharwal
Written by Thomas (Tingkai) Liu and Yoseph Maguire

This code is designed to process raw csv files output from accelerometers
attached to a bike pedal, and output the overall cadence of the rider over
the dataset.
It needs to be updated to log cadence over time using SFFTs. 

"""

# Possibly use scipy.signal
"""
Just for speed testing purposes: be sure to include Bike_4XYZ.csv.
"""




def bikedata(fname):

    sMat = retrieveacc("{}".format(fname))
    # User Acceleration
    ux_x = sMat['user_acc_x']
    ux_y = sMat['user_acc_y']
    ux_z = sMat['user_acc_z']
    # ux_x = sMat['user_acc_x(G)']
    # ux_y = sMat['user_acc_y(G)']
    # ux_z = sMat['user_acc_z(G)']
    ux = [ux_x,ux_y,ux_z]

    X_Mean = float(np.mean(ux_x))
    Y_Mean = float(np.mean(ux_y))
    Z_Mean = float(np.mean(ux_z))
    pow = [np.linalg.norm(ux_x),np.linalg.norm(ux_y),np.linalg.norm(ux_z)]
    # data = ux[pow.index(max(pow))] # set data as the axis with the largest power.
    data = ux_x
    # data = np.sqrt(np.square(X_accel)+np.square(Y_accel)+np.square(Z_accel)) # Use if worried about orientation.
    sxx = stft(data)
    Fs = 100
    var = np.mean((data-np.mean(data))**2) 

    plotstft(sxx)

    
    dlen = len(data)
    np2 = nextpow2(dlen)
    fftlength = np2
    ctr = int((fftlength/2))
    print(ctr)
    faxis = np.multiply(Fs/2,np.linspace(0,1,ctr))
    b, a = signal.butter(4, [.01, .5], 'bandpass', analog=False)
    filt_d = signal.lfilter(b, a, data)
    fild_d = data
    fdata = np.fft.fft(filt_d,fftlength) #/ len(data) # Possibly need to normalize by the length of data.
    mag = abs(fdata[0:ctr])
    print(fdata[0:ctr])
    # fig = plt.figure()
    # ax = fig.add_subplot(1,1,1)
    # ax.plot(faxis,mag, 'b-',linewidth=2,label=r'$y=\sin(x)$')
    # ax.set_ylabel(r'$y$',fontsize=40)
    # ax.set_xlabel(r'$x$',fontsize=40)
    # ax.legend(loc='best',fontsize=40)
    # ax.grid(True)
    # fig.suptitle(r'$The\ Frequency\ Domain$',fontsize=40)
    # # fig.tight_layout(pad=0)
    plt.show()
    # fig.savefig('filename.png',dpi=125)



    font = {'family': 'sans-serif',
            'weight': 'bold',
            'size': 40}


    rc('font', **font)
    # plt.plot(faxis, mag)
    # plt.title("Fourier Transform")
    # plt.xlabel("Frequency")
    # plt.ylabel("Amplitude")
    # plt.show()
    mag_max = 0
    mag_idx = 0
    for i in range(0,len(mag)):
        if mag[i] > mag_max:
            mag_max = mag[i]
            mag_idx = i/ctr*(Fs/2)
    rpm = mag_idx * 60

    # winlen = len(data)//10
    # np2w = nextpow2(winlen)

    # f, t, Sxx = signal.spectrogram(data, fs=100, window='hamming', nperseg=winlen, noverlap=winlen//4, nfft=np2w,
    #                            detrend='constant', return_onesided=True, scaling='density')
    # plt.pcolormesh(t, f, Sxx)
    # plt.ylabel('Frequency [Hz]')
    # plt.xlabel('Time [sec]')
    # plt.show()
    # print("Primary Sxx:",len(Sxx))
    # print("Nested Sxx:", Sxx[0])
    # print("Frequency:", len(f))
    # print("Time:", len(t))
    # peaks = []
    # # peaks = [max(row[i]) for i in range(len(Sxx[0])) for row in Sxx]
    # for i in range(len(Sxx[0])):
    #     cp = 0
    #     for row in Sxx:

    #         if cp < row[i]:
    #             cp = row[i]
    #     peaks.append(cp*60)

    # tsxx = np.transpose(Sxx)
    # print("Primary Sxx:", len(tsxx))
    # print("Nested Sxx:", len(tsxx[0]))
    # peaks = [(60*max(row)) for row in tsxx]
    # print("Average RPM from Peaks:", mean(peaks))
    # return rpm, peaks

    return rpm, 1

def stft(data,Fs=100):
    """
    Input: 3 Axis Data
    Output: Power Representation of
    """
    # Remove the DC Component of the function
    data = noDC(data)
    wlen = Fs * 10 # Window length = 10
    segs = len(data) // wlen # integer divsion to return number of segments of data
    windsegs = []
    numz = nextpow2(wlen) - wlen
    j = 0
    win = []
    for i in data:
        if j < wlen:  # append the value from j=0 to j = (number of samples - 1)
            win.append(i)
        else:
            j = 0
            for i in range(0, numz):  # Zero padding the window
                win.append(0)
            windsegs.append(win)  # Add that window to the segmented dataset
            win = [i]  # Reset the window variable
        j += 1
    b, a = signal.butter(4, [.01, .5], 'bandpass')
    dft=[]
    winlen = int(len(windsegs[0]))
    for seg in windsegs:
        seg = signal.lfilter(b, a, seg)
        wind = signal.get_window(('kaiser', 4.0), winlen)  # Beta Parameter of Kaiser: 4. Num Samples in window: 9.
        snip = seg * wind
        nfft = nextpow2(wlen)
        A = np.fft.fft(snip, nfft)
        dft.append(A)
    return dft

def noDC(data):
    D_mean = float(np.mean(data))
    for i in range(0, len(data)):
        data[i] -= D_mean
    return data

def plotstft(sxx, Fs=100):
    winlen = int(len(sxx[0]))

    # with plt.():
    fig1 = plt.figure()
    ax = fig1.add_subplot(1,1,1)
    ctr = int(winlen / 2)
    faxis = np.multiply(Fs / 2, np.linspace(0, 1, ctr))*60
    ratio = []
    for dft in sxx:
        mag = abs(dft[0:ctr])
        max_idx, max_val = max(enumerate(mag), key = lambda p: p[1])
        ptotal = np.square(np.linalg.norm(mag,2))
        pmax = np.square(np.absolute(max_val))
        # print('max power: {}'.format(max_val))
        frac = pmax/ptotal
        # print(frac)
        ratio.append(frac)

        ax.plot(faxis, mag, linewidth=3.0)
        ax.set_xlabel('Frequency (RPM)')
        ax.set_ylabel('|H(f)|')
        ax.set_title('User 4 STFT Spectrum')
        font = {'family' : 'sans-serif ',
                'weight' : 'bold',
                'size'   : 30}
        rc('font', **font)

    fig1.savefig('STFTPlot.png')
    fig2 = plt.figure()
    plt.stem(np.linspace(1,len(ratio),num=len(ratio)),ratio, linewidth=3.0)
    plt.xlabel('Window (10s)', fontsize = 60)
    plt.ylabel('Symmetry in Pedaling', fontsize = 60)
    fig2.savefig('RatioPlot.png')
        # print("Ratio:{}".format(ratio))



def nextpow2(n):
        """
        n = integer.
        Bike_1.csv = 38558
        Return: Next largest value that is equal to 2^x
        """

        n -= 1
        n |= n >> 1
        n |= n >> 2   
        n |= n >> 4
        n |= n >> 8
        n |= n >> 16  
        n += 1  
        return n

def retrieveacc(fname):
    sMat = {} # Storage Matrix is a dictionary I suppose
    fpath = os.path.join(os.path.dirname(__file__),fname)
    with open(fpath,mode='r',newline=None) as csvfile:

        parsed = csv.reader(csvfile)
        parsed = np.array(list(parsed))
        titles, parsed = parsed[0], parsed[1:]
        data = parsed.transpose()
        data = cleandata(data[0:16]) # 16 columns because the other ones aren't useful.
        cidx = 0
        for col in data:
            sMat[titles[cidx]] = col
            cidx += 1
    return sMat


def quicksort(tarr):
    """
    Input must be a enumerated list with the index in [0] and value in [1]
    """
    if len(tarr) <= 1:
        return tarr
    else:
        l, pivot, r = partition(tarr)
        return quicksort(l)+[pivot]+quicksort(r)

def partition(tarr):
    pivot, tarr = tarr[len(tarr)-1], tarr[:len(tarr)-1]
    l = [x for x in tarr if x[1] <= pivot[1]]
    r = [x for x in tarr if x[1] > pivot[1]]
    return l, pivot, r

def cleandata(data):
    """
    Data needs to be represented as columns.
    Assumes uniform breaks, so that removing all empty values
    will uniformly shrink the data.
    """
    parsed = np.array([[float(x) for x in col if x != ''] for col in data])
    return parsed

def main():
    try:
        filename = sys.argv[1]
    except IndexError:
        print("Did not enter a filename.")
    else:
        rpm, stfts = bikedata(filename)
        print("Average RPM over Data:", str(rpm))
        print("Peaks:", str(stfts))



main()

File: Bikedata.m
function Bikedata(filename) 

close all;

Fs = 100; % Visually inspecting the oscilloscope
%filename = 'Bike_1.csv'
SENS_DATA = importdata(filename);
parsedata = SENS_DATA.data;
dlen = length(parsedata(:,1));
fdel = 20;
edel = 20;
SENS_DATA = parsedata(fdel*Fs:dlen-edel*Fs,:);
% SENS_DATA = parsedata(1:timedel*Fs,:);
timestamp = SENS_DATA(:,1);
% ATT_R = SENS_DATA(:,2);
% ATT_P = SENS_DATA(:,3);
% ATT_Y = SENS_DATA(:,4);
% ROTR_X = SENS_DATA(:,5);
% ROTR_Y = SENS_DATA(:,6);
% ROTR_Z = SENS_DATA(:,7);
% GR_X = SENS_DATA(:,8);
% GR_Y = SENS_DATA(:,9);
% GR_Z = SENS_DATA(:,10);
UA_X = SENS_DATA(:,11);
UA_Y = SENS_DATA(:,12);
UA_Z = SENS_DATA(:,13);
% MAG_X = SENS_DATA(:,12);
% MAG_Y = SENS_DATA(:,13);
% MAG_Z = SENS_DATA(:,14);


keep_rows = find(~isnan(UA_X)); % find the 1s produced by isnan.

% User acceleration
Accel_X = UA_X(keep_rows);
Accel_Y = UA_Y(keep_rows);
Accel_Z = UA_Z(keep_rows);
ts = timestamp(keep_rows);
ts = ts - ts(1);
[~,sorted] = sort(ts);
Accel_X = Accel_X(sorted);
% Accel_X(:) = 0;
Accel_Y = Accel_Y(sorted);
% Accel_Y(:) = 0;
Accel_Z = Accel_Z(sorted);
% Accel_Z(:) = 0;
% accel = [Accel_X Accel_Y Accel_Z];
ws = 40
we = 50
Accel_X = Accel_X(ws*Fs:we*Fs);
Accel_Y = Accel_Y(ws*Fs:we*Fs);
Accel_Z = Accel_Z(ws*Fs:we*Fs);

% Time length of signal

sigtime = length(Accel_X)/Fs/60.0;
writing = sprintf('\nThe signal length is %d minutes.\n',sigtime); 
disp(writing);


powX = bandpower(Accel_X); % Find the power contained between 30 and 40s of Acceleration data after removing 20s noise.
powY = bandpower(Accel_Y);
powZ = bandpower(Accel_Z);
powers = [powX powY powZ];
ptot = sum(powers);
[~,powidx] = max(powers);

% npowX = bandpower(Accel_X(1:11*Fs));
% npowY = bandpower(Accel_Y(1:11*Fs));
% npowZ = bandpower(Accel_Z(1:11*Fs));

% % Remove the signal with the least power.
% % This may be redundant considering weighting.
% 
% switch powidx
%     case 1
%         
%         Accel_X(:) = 0;
%         disp('X has lowest power')
%     case 2
%         Accel_Y(:) = 0;
%         powY = 0;
%         disp('Y has lowest power')
%     case 3
%         Accel_Z(:) = 0;
%         disp('Z has lowest power')
% end
%     

switch powidx % for using with max(powers)
    case 1
        data = Accel_X;
    case 2
        data = Accel_Y;
    case 3
        data = Accel_Z;
end




X_weight = 3*powX/ptot;
Y_weight = 3*powY/ptot;
Z_weight = 3*powZ/ptot;

t = linspace(0,length(Accel_X)/Fs,length(Accel_X));




figure;
subplot(211);
% plot(t,Accel_X,'b');
hold on;
% plot(t,Accel_Y,'m');
% plot(t,Accel_Z,'r');

xlabel('Time (sec)', 'FontSize' ,10);
ylabel('User Acceleration (m/s^2)', 'FontSize', 10);
title('Cadence of Bike User','FontSize',10);


%data = sqrt((Z_weight.*Accel_Z).^2+(Y_weight.*Accel_Y).^2+(X_weight.*Accel_X).^2);
data = Accel_X;
data = data-mean(data);% subtract DC value
plot(t,data,'b','linewidth',2); 
% findpeaks(data, Fs,'MinPeakDistance',1/100,'MinPeakHeight',0);
hold off;
% legend('X','Y','Z','Selected');



% Unfiltered Data
fftlength = 2^nextpow2(length(data));
L = length(data);
fdata = fft(data, fftlength) / L;
ctr = (fftlength / 2) + 1;
faxis = 60*(Fs / 2) .* linspace(0,1, ctr); % multiply by 60 for RPM vs RPS
mag = abs(fdata(1:ctr));
disp('Maximum Index:');
[~,idx] = max(mag)
%fftcdnc = faxis(idx)

% Plot Unfiltered Data
subplot(212);
plot(faxis,mag,'LineWidth',2); hold on;
% findpeaks(mag,faxis,'MinPeakHeight',.005)
[pkt,lct] = findpeaks(mag,faxis,'MinPeakHeight',.005);
lct(1:5)
% plot(lct,pkt,'x');
[spkt,slct] = sort(pkt,'descend');
xsorted = lct(slct);
maxVal = spkt(1:5)
maxIdx = xsorted(1:5)
% plot(maxIdx,maxVal,'x');
B = mag(slct);


title('FFT of Weighted Averaged Data','FontSize' ,10);
xlabel('Frequency (RPM)','FontSize' ,10);
ylabel('Magnitude','FontSize' ,10);


% Filtered Data
cutfreq = [.25 15];
[b,a] = butter(4,cutfreq./(Fs./2));
lpf = filter(b,a,data)  ;
bdata = fft(lpf,fftlength) / L;
[~,idx] = max(abs(bdata(1:ctr)));
% bdata(idx) = 0;
fftcadence = faxis(idx)
powah = bandpower(lpf,Fs,[fftcadence/60-.05 fftcadence/60+.05])
totpowah = bandpower(lpf)

plot(faxis,abs(bdata(1:ctr)),'linewidth',1);
legend('unfilt','butter LPF'); hold off;


    
% figure;
% subplot 221;
    % windlen = floor(length(lpf)/10);
    windt = 4;
    windlen = floor(windt*Fs);
    nlap= [];
    nfft=2^nextpow2(windlen);
    wind = hamming(windlen);
%     spectrogram(lpf,wind,nlap,nfft,Fs,'yaxis');hold on;
%     title('Spectrogram (4s)');
%     % colorbar;
%     %plot(t,f(I),q,'r','linewidth',2); 
%     hold off;
%     
% 
% 
% subplot 222;
    [s,f,t,pxx]=spectrogram(lpf,wind,nlap,nfft,Fs,'yaxis');
%     disp(['Spectrogram time: ', num2str(t(length(t))),' seconds.']);
    [~,I] = max(10*log10(pxx)); % largest PSD in each column (STFT).
    cadot = 60*f(I); %cadence over time
%     cadlen = length(cadot);
%     x = linspace(0,t(length(t)),cadlen);
%     stairs(x,cadot,'linewidth',2);
%     grid on;
%     title('Discrete Cadence Over Time');
%     xlabel('Time (mins)');
%     ylabel('Cadence (RPM)');
% 
% subplot 223;
%     windt = 12.8;
%     windlen = floor(windt*Fs);
%     nlap= [];
%     nfft=2^nextpow2(windlen);
%     wind = hamming(windlen);
%     spectrogram(lpf,wind,nlap,nfft,Fs,'yaxis');hold on;
%     title('Spectrogram (12.8s)');
% 
% subplot 224;
    [~,f,t,pxx]=spectrogram(lpf,wind,nlap,nfft,Fs,'yaxis');
    [M,I] = max(10*log10(pxx)); % largest PSD in each column (STFT).
    M
    cadot = 60*f(I);
    cadlen = length(cadot);
    x = linspace(0,t(length(t)),cadlen);
%     stairs(x,cadot,'linewidth',2);
%     grid on;
%     title('Discrete Cadence Over Time');
%     xlabel('Time (mins)');
%     ylabel('Cadence (RPM)');









avgcadence = mean(cadot);

% save output over time to .csv file
fname = strsplit(filename,'.csv');
newfilename = sprintf('%speaks.csv',fname{1});
q = [x',cadot];
disp(['Window Length: ',num2str(windlen)]);
disp(['FFT Peak: ',num2str(fftcadence)]);
disp(['STFT Avg: ', num2str(avgcadence)]);
csvwrite(newfilename,q);


end
